src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/ReceiveMessagesEffect.kt

```diff
@@ -0,0 +1,50 @@
+package com.pubnub.api.subscribe.eventengine.effect
+
+import com.pubnub.api.PubNubException
+import com.pubnub.api.endpoints.remoteaction.RemoteAction
+import com.pubnub.api.eventengine.EventDeliver
+import com.pubnub.api.eventengine.ManagedEffect
+import com.pubnub.api.subscribe.eventengine.event.Event
+
+class ReceiveMessagesEffect(
+    private val receiveMessagesProvider: ReceiveMessagesProvider,
+    private val eventDeliver: EventDeliver,
+    private val receiveMessagesInvocation: SubscribeEffectInvocation.ReceiveMessages
+) : ManagedEffect {
+    override fun runEffect(completionBlock: () -> Unit) {
+        val receiveMessagesRemoteAction = getReceiveMessagesRemoteAction()
+        receiveMessagesRemoteAction.async { result, status ->
+            try {
+                if (status.error) {
+                    eventDeliver.passEventForHandling(
+                        Event.ReceiveFailure(
+                            status.exception
+                                ?: PubNubException("Unknown error") // todo check it that can happen
+                        )
+                    )
+                } else {
+                    eventDeliver.passEventForHandling(
+                        Event.ReceiveSuccess(
+                            result!!.messages,
+                            result.subscriptionCursor
+                        )
+                    )
+                }
+            } finally {
+                completionBlock()
+            }
+        }
+    }
+
+    override fun cancel() {
+        val receiveMessagesRemoteAction = getReceiveMessagesRemoteAction()
+        receiveMessagesRemoteAction.silentCancel()
+    }
+
+    private fun getReceiveMessagesRemoteAction(): RemoteAction<ReceiveMessagesResult> =
+        receiveMessagesProvider.receiveMessages(
+            receiveMessagesInvocation.channels,
+            receiveMessagesInvocation.channelGroups,
+            receiveMessagesInvocation.subscriptionCursor
+        )
+}
```

Comment:
You could change this class so it depends only fewer things. It will make the tests of it simpler, and the implementation is shorter as well. Also it fixes the bug mentioned by me already in other comment.

```suggestion
class ReceiveMessagesEffect(
    private val remoteAction: RemoteAction<ReceiveMessagesResult>,
    private val eventDeliver: EventDeliver
) : ManagedEffect {
    override fun runEffect(completionBlock: () -> Unit) {
        remoteAction.async { result, status ->
            try {
                if (status.error) {
                    eventDeliver.passEventForHandling(
                        Event.ReceiveFailure(
                            status.exception
                                ?: PubNubException("Unknown error") // todo check it that can happen
                        )
                    )
                } else {
                    eventDeliver.passEventForHandling(
                        Event.ReceiveSuccess(
                            result!!.messages,
                            result.subscriptionCursor
                        )
                    )
                }
            } finally {
                completionBlock()
            }
        }
    }

    override fun cancel() {
        val receiveMessagesRemoteAction = remoteAction
        receiveMessagesRemoteAction.silentCancel()
    }
}

```